course/api/views.py

```diff
@@ -64,6 +95,112 @@ def get_queryset(self):
     def get_object(self):
         return self.get_member_object('instance', 'Course')
 
+    def set_teachers(self, course: CourseInstance, data: dict) -> None:
+        teachers = []
+        for i in data['teachers']:
+            userprofile: UserProfile = None
+            try:
+                user = User.objects.get(username=i)
+            except User.DoesNotExist:
+                # If user does not exist, create a new user.
+                # If external authentication (e.g. Shibboleth) is used, other
+                # attributes will be updated when user logs in for the first time.
+                user = User.objects.create_user(i)
+
+            userprofile = UserProfile.objects.get(user=user)
+            teachers.append(userprofile)
+
+        course.set_teachers(teachers)
+
+    def create(self, request: HttpRequest, version) -> Response:
+        if (not request.user.is_superuser):
+            return Response(
+                'Only admins can add new course instance.',
+                status=status.HTTP_403_FORBIDDEN
+            )
+
+        data = request.data
+        form = CourseInstanceForm(data)
+
+        if not form.is_valid():
+            return Response({'status': 'error', 'errors': form.errors}, status=status.HTTP_400_BAD_REQUEST)
+
+        else:
+            code = form.cleaned_data.get('code')
+            course: Course = Course.objects.filter(code=code).first()
+            if not course:
+                try:
+                    course = Course.objects.create(
+                        name = form.cleaned_data.get('name'),
+                        code = code,
+                        url = form.cleaned_data.get('course_url'),
+                    )
+                except:
+                    return Response('Course creation failed.',
+                                    status=status.HTTP_400_BAD_REQUEST)
+
+            formdata = form.cleaned_data
+            instance_name = formdata.get('instance_name')
+            if (not formdata.get('url')):
+                # if url is not given, compose it from instance name, removing special chars
+                formdata['url'] = ''.join(i for i in instance_name if i.isalnum())
+            
+            try:
+                new_instance: CourseInstance = CourseInstance.objects.create(
+                    instance_name = instance_name,
+                    url = formdata.get('url'),
+                    language = formdata.get('language'),
```

Comment:
What happens if the request data excludes some parameters? E.g., language may be blank in the model validation, but does it allow assigning the `None` value here? Assigning the empty string "" when the value is really empty would be more correct concerning the field types.